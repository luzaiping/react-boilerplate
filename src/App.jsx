import React from 'react';
import axios from 'axios';

import Counter from './Counter';

export const dataReducer = (state, action) => {
  const { type, list } = action;
  if (type === 'SET_ERROR') {
    return { ...state, list: [], error: true };
  }

  if (type === 'SET_LIST') {
    return { ...state, list, error: null };
  }

  throw new Error();
};

const initialData = {
  list: [],
  error: null
};

const App = () => {
  const [count, setCount] = React.useState(0);
  const [data, dispatch] = React.useReducer(dataReducer, initialData);

  React.useEffect(() => {
    axios
      .get('http://hn.algolia.com/api/v1/search?query=react')
      .then(response => {
        dispatch({ type: 'SET_LIST', list: response.data.hits });
      })
      .catch(() => {
        dispatch({ type: 'SET_ERROR' });
      });
  }, []);

  return (
    <div>
      <h1>My Counter</h1>
      <Counter count={count} />

      <button type="button" onClick={() => setCount(count + 1)}>
        Increment
      </button>

      <button type="button" onClick={() => setCount(count - 1)}>
        Decrement
      </button>

      <h2>My Async Data</h2>
      {data.error && <div className="error">Error</div>}
      <ul>
        {data.list.map(item => (
          <li key={item.objectID}>{item.title}</li>
        ))}
      </ul>
    </div>
  );
};

export default App;
