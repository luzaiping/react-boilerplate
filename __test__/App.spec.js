import React from 'react';
import renderer from 'react-test-renderer';

import App, { dataReducer } from '../src/App';

const list = ['a', 'b', 'c'];

describe('App', () => {
  test('snapshot render', () => {
    const component = renderer.create(<App />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('Reducer', () => {
    it('should set a list', () => {
      const state = { list: [], error: null };
      const newState = dataReducer(state, {
        type: 'SET_LIST',
        list
      });
      expect(newState).toEqual({ list, error: null });
    });

    test('should reset the error if list is set', () => {
      const state = { list: [], error: true };
      const newState = dataReducer(state, {
        type: 'SET_LIST',
        list
      });
      expect(newState).toEqual({ list, error: null });
    });

    test('should set the error', () => {
      const state = { list: [], error: null };
      const newState = dataReducer(state, {
        type: 'SET_ERROR'
      });
      expect(newState).toEqual({ list: [], error: true });
      expect(newState.error).toBeTruthy();
    });
  });
});
