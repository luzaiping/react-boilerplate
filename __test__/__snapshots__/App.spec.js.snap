// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot render 1`] = `
<div>
  <h1>
    My Counter
  </h1>
  <div>
    <p>
      0
    </p>
  </div>
  <button
    onClick={[Function]}
    type="button"
  >
    Increment
  </button>
  <button
    onClick={[Function]}
    type="button"
  >
    Decrement
  </button>
  <h2>
    My Async Data
  </h2>
  <ul />
</div>
`;
